### PR TITLE
Rust: Basic data flow through arrays

### PR DESCRIPTION
**Pull request by paldepind**
Adds basic data flow through arrays. Basic in the sense that we don't keep track of element positions.

### CODE REVIEW

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -623,6 +623,15 @@ private class StructFieldContent extends Content, TStructFieldContent {
   override string toString() { result = s.toString() + "." + field_.toString() }
 }
 
+/**
+ * Content stored at an element in an array.
+ */
+final private class ArrayElementContent extends VariantContent, TArrayElement {
```
Comment (hvitved) [inline rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:629]:
```suggestion
final private class ArrayElementContent extends Content, TArrayElement {
```

**Review by hvitved — CHANGES_REQUESTED** [review]
LGTM, just one small thing. We should also implement [`defaultImplicitTaintRead`](https://github.com/github/codeql/blob/c4e53b842d66e3bd1ce53e4e5d63afc67a0a17e0/ruby/ql/lib/codeql/ruby/dataflow/internal/TaintTrackingPrivate.qll#L21) to allow for implicit array reads at sinks, and also [array reads as local taint steps](https://github.com/github/codeql/blob/c4e53b842d66e3bd1ce53e4e5d63afc67a0a17e0/ruby/ql/lib/codeql/ruby/dataflow/internal/TaintTrackingPrivate.qll#L117-L127). We can do both follow-up, if you prefer.

**Comment by paldepind** [comment]
I see. I'll do it in this PR once #18202 is merge.

File: rust/ql/lib/codeql/rust/dataflow/internal/TaintTrackingImpl.qll [hunk 2]
```diff
@@ -35,6 +34,15 @@ module RustTaintTracking implements InputSig<Location, RustDataFlow> {
...
         succ.asExpr() = index
       )
+      or
+      // Although data flow through collections is modeled using stores/reads,
+      // we also allow taint to flow out of a tainted collection. This is
+      // needed in order to support taint-tracking configurations where the
+      // source is a collection.
+      exists(ContentSet cs |
+        RustDataFlow::readStep(pred, cs, succ) and
+        cs.(SingletonContentSet).getContent() instanceof ArrayElementContent
```
Comment (hvitved) [inline rust/ql/lib/codeql/rust/dataflow/internal/TaintTrackingImpl.qll:44]:
```suggestion
      exists(SingletonContentSet cs |
        RustDataFlow::readStep(pred, cs, succ) and
        cs.getContent() instanceof ArrayElementContent
```

**Review by hvitved — CHANGES_REQUESTED** [review]

**Review by hvitved — APPROVED** [review]

**Pull request merged**